lambda/shared/src/main/scala/feral/lambda/tracing/TracedLambda.scala

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2021 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package feral.lambda.tracing
+
+import cats._
+import cats.data._
+import cats.effect.std.Random
+import cats.effect.{Trace => _, _}
+import natchez._
+import feral.lambda._
+import natchez.noop.NoopSpan
+import natchez.xray.{XRay, XRayEnvironment}
+
+abstract class TracedLambda[F[_]: MonadCancelThrow, G[_], Event] {
+  def extractKernel(event: Event, context: Context[F]): F[Kernel]
+
+  def runTraced[A](span: Span[F])(ga: G[A]): F[A]
```

Comment:
Same with this one, we're playing with some ideas in https://github.com/tpolecat/natchez/pull/448.

Reply:
Sounds good, I pushed up a version using an inlined copy of `LiftTrace` for now, while that's discussed over there.

Reply:
I updated the LiftTrace concept with some new ideas, check the PR. I'm about to push a branch trying out that one instead.